jenkins/http-parser.jenkinsfile

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env groovy
+
+import groovy.json.JsonOutput
+
+pipeline {
+    agent any
+
+    parameters {
+        booleanParam(name: 'CERTIFY_SAFE', defaultValue: false, description: 'I have reviewed *the latest version of* these changes and I am sure that they don’t contain any code that could compromise the security of the CI infrastructure.')
+        string(defaultValue: "", description: 'The numeric ID of the pull request (from GitHub URL)', name: 'PR_ID')
+    }
+
+    stages {
+        stage('Clone repository') {
+            steps {
+                git 'https://github.com/nodejs/http-parser.git'
```

Comment:
When I try the pipeline script generator to check out a PR branch of this repo, it suggests:

```groovy
git branch: 'pull/${env.PR_ID}/head', url: 'https://github.com/nodejs/http-parser'
```

Could you use that rather than this and tne next stage?

Reply:
The reason why I made the PR-branch-switching a separate stage was so CI runs could be done for master... though maybe this is unneeded functionality right now? Might be better to start off with just PRs? 😬

Reply:
Ahh, good point. Then alternately just do:

```groovy
git(branch: '${env.BRANCH', url: 'https://github.com/nodejs/http-parser')
```

And put `BRANCH` as a parameter:

```groovy
         string(name: 'BRANCH', 
                   defaultValue: "refs/pull/XXXX/head", 
                   description: 'Branch to build, e.g. "master" or "refs/pull/XXXX/head" where XXXX is the PR number') 
```

Reply:
Hmm, trying to get this working on my local instance, but here's the part that's failing:

```console
[Pipeline] git
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/nodejs/http-parser.git # timeout=10
Fetching upstream changes from https://github.com/nodejs/http-parser.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/nodejs/http-parser.git +refs/heads/*:refs/remotes/origin/*
 > git rev-parse refs/remotes/origin/refs/pull/388/head^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/origin/refs/pull/388/head^{commit} # timeout=10
 > git rev-parse origin/refs/pull/388/head^{commit} # timeout=10
```

I have a feeling that `+refs/heads/*:refs/remotes/origin/*` is the issue, but I haven't been able to find a way to change that 😞

Reply:
@maclover7 I can't see the error there, what exactly doesn't work?

Reply:
Sorry, my bad for just pasting in an error, here's [a gist](https://gist.github.com/maclover7/21693d452114fd51f27002013772321a) with the updated pipeline file and the full job log. Seems like specifying the branch via `git` is a little wonky :(

Reply:
FYI, I was playing around with this for a `node-inspect` build script and came up with the following:

```groovy
        stage("Checkout") {
            steps {
                checkout(poll: false, scm: [
                    $class: 'GitSCM',
                    branches: [[
                        name: (params.isPR ? "pr/${params.projectRef}/head" : params.projectRef)
                    ]],
                    userRemoteConfigs: [[
                        url: "https://github.com/${params.projectRepo}",
                        refspec: '+refs/pull/*:refs/remotes/origin/pr/* +refs/heads/*:refs/remotes/origin/*',
                    ]],
                ])
            }
        }
```